### Build a DNN using Keras with `RELU` and `ADAM`

#### Load tensorflow

In [1]:
import tensorflow as tf

#### Collect Fashion mnist data from tf.keras.datasets 

In [2]:
from keras.datasets import fashion_mnist
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

Using TensorFlow backend.


4423680/4422102 [==============================] - 1s 0us/step


#### Change train and test labels into one-hot vectors

In [0]:
from keras.utils import to_categorical

y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

#### Build the Graph

#### Initialize model, reshape & normalize data

In [0]:
from keras.models import Sequential
from keras.layers import Dense, Activation, Dropout, Flatten, Reshape
from keras.layers import Convolution2D, MaxPooling2D

In [0]:
# Clear out tensorflow memory

tf.keras.backend.clear_session()

In [6]:
# Initializing the Sequential model

model = Sequential()

In [0]:
# Normalizing the data

x_train = x_train.astype('float32') / 255
x_test = x_test.astype('float32') / 255

In [0]:
# Reshaping the data

x_train = x_train.reshape((x_train.shape[0], 28, 28, 1)).astype('float32')
x_test = x_test.reshape((x_test.shape[0], 28, 28, 1)).astype('float32')

#### Add two fully connected layers with 200 and 100 neurons respectively with `relu` activations. Add a dropout layer with `p=0.25`

In [9]:
# 1st Conv Layer
model.add(Convolution2D(32, 3, 3, input_shape=(28, 28, 1)))
model.add(Activation('relu'))

# 2nd Conv Layer
model.add(Convolution2D(32, 3, 3))
model.add(Activation('relu'))

model.add(Flatten())

# 1st hidden layer
model.add(Dense(200, activation='relu'))

# 2nd hidden layer
model.add(Dense(100, activation='relu'))

# Dropout layer
model.add(Dropout(0.25))




Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: UserWarning: Update your `Conv2D` call to the Keras 2 API: `Conv2D(32, (3, 3), input_shape=(28, 28, 1...)`
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:5: UserWarning: Update your `Conv2D` call to the Keras 2 API: `Conv2D(32, (3, 3))`
  """


### Add the output layer with a fully connected layer with 10 neurons with `softmax` activation. Use `categorical_crossentropy` loss and `adam` optimizer and train the network. And, report the final validation.

In [10]:
# Output layer
model.add(Dense(10, activation='softmax'))

# Loss and Optimizer
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# Storing Training Results
from keras.callbacks import EarlyStopping
early_stopping = EarlyStopping(monitor='val_acc', patience=5, verbose=1, mode='auto')
callback_list = [early_stopping]

# Training the model
model.fit(x_train, y_train, batch_size=32, nb_epoch=10, validation_data=(x_test, y_test), callbacks=callback_list)



Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:12: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  if sys.path[0] == '':




Train on 60000 samples, validate on 10000 samples
Epoch 1/10





60000/60000 [==============================] - 18s 304us/step - loss: 0.4093 - acc: 0.8539 - val_loss: 0.3064 - val_acc: 0.8861
Epoch 2/10
60000/60000 [==============================] - 11s 184us/step - loss: 0.2508 - acc: 0.9078 - val_loss: 0.2480 - val_acc: 0.9077
Epoch 3/10
60000/60000 [==============================] - 11s 183us/step - loss: 0.1867 - acc: 0.9305 - val_loss: 0.2629 - val_acc: 0.9080
Epoch 4/10
60000/60000 [==============================] - 11s 182us/step - loss: 0.1361 - acc: 0.9495 - val_loss: 0.2619 - val_acc: 0.9138
Epoch 5/10
60000/60000 [==============================] - 11s 183us/step - loss: 0.0982 - acc: 0.9635 - val_loss: 0.3129 - val_acc: 0.9115
Epoch 6/10
60000/60000 [==============================] - 11s 181us/step - loss: 0.0675 - acc: 0.9750 - val_loss: 0.3307 - val_acc: 0.9162
Epoch 7/10
60000/60000 [==============================] - 11s 181us/step - loss: 0.0519 - acc: 0.9807 - val_l

In [11]:
score1 = model.evaluate(x_train, y_train)
score2 = model.evaluate(x_test, y_test)
print('\nTrain accuracy:', score1[1], '\nTest accuracy:', score2[1])

10000/10000 [==============================] - 0s 49us/step

Train accuracy: 0.9952833333333333 
Test accuracy: 0.9119


#### Model Summary: 
The model identified different fashion products from the given images using a CNN model. The 'dropout’ technique was used for this problem. The train accuracy is 99.5% and the test accuracy is 91.2%, which gives us the idea about no overfitting being done. Also, there may be a chance to improve performance further, by increasing the number of convolutional layers (and neurons/filters) or trying out different combinations of different layers.